plugins/dell/fu-plugin-dell.c

```diff
@@ -522,6 +523,114 @@ fu_plugin_get_results (FuPlugin *plugin, FuDevice *device, GError **error)
 	return TRUE;
 }
 
+static void Esys_Finalize_autoptr_cleanup (ESYS_CONTEXT *esys_context)
+{
+	Esys_Finalize (&esys_context);
+}
+G_DEFINE_AUTOPTR_CLEANUP_FUNC (ESYS_CONTEXT, Esys_Finalize_autoptr_cleanup)
+
+static gchar *
+fu_plugin_dell_get_tpm_capabilitity (ESYS_CONTEXT *ctx, guint32 query, GError **error)
+{
+	TSS2_RC rc;
+	guint32 val;
+	g_autofree gchar *result = NULL;
+	g_autofree TPMS_CAPABILITY_DATA *capability = NULL;
+	rc = Esys_GetCapability (ctx, ESYS_TR_NONE, ESYS_TR_NONE, ESYS_TR_NONE,
+	                         TPM2_CAP_TPM_PROPERTIES, query, 1, NULL, &capability);
```

Comment:
If the requested property `query` does not exist, the TPM returns the next existing property instead, see the [TPM 2.0 specification Part 3, Section 30.2.1](https://trustedcomputinggroup.org/wp-content/uploads/TPM-Rev-2.0-Part-3-Commands-01.38.pdf#page=354):

> The *capability* parameter determines the category of data returned. The *property* parameter selects the first value of the selected category to be returned. If there is no property that corresponds to the value of *property*, the next higher value is returned, if it exists.

It would therefore be safer to additionally check if
- `data.tpmProperties.count > 0` (has any property been returned at all?) and
- `data.tpmProperties.tpmProperty[0].property == query` (is the returned property the one we requested?).

These shouldn't be hard errors though, returning NULL would be enough in this case.

Reply:
Thanks.  I'll make some adjustments for this.